### PR TITLE
Fixes Issue #6321 caused when proxy.config.http.no_dns_just_forward_to_parent

### LINKED ISSUE #6321: strict round robin parent select doesn't work as expected
We've identified two issues regarding strict parent proxy select policy.
Our parent.config looks like this:
```
dest_domain=. parent="10.132.0.112:3120,10.132.0.112:3121,10.132.0.112:3122,10.132.0.112:3123,10.132.0.112:3124,10.132.0.112:3125,10.132.0.112:3126,10.132.0.112:3127" parent_is_proxy=false round_robin=strict
```

on long-lived ATS instances ss shows that all the traffic is being handled by the first parent. This seems to be related to the fact that rr_next (declared as uint32_t in https://github.com/apache/trafficserver/blob/master/proxy/ParentSelection.h#L143) gets casted as int32_t in https://github.com/apache/trafficserver/blob/master/proxy/ParentRoundRobin.cc#L102 and as soon as the value overflows, cur_index always begins on a negative value, returning always the parent with index 0:
```
Jan 14 17:40:41 cp4027 traffic_manager[20323]: [Jan 14 17:40:41.847] {0x2ade5d766700} DEBUG: <ParentRoundRobin.cc:61 (selectParent)> (parent_select) In ParentRoundRobin::selectParent(): Using a round robin parent selection strategy.
Jan 14 17:40:41 cp4027 traffic_manager[20323]: [Jan 14 17:40:41.847] {0x2ade5d766700} DEBUG: <ParentRoundRobin.cc:140 (selectParent)> (parent_select) cur_index: -4, result->start_parent: -4
Jan 14 17:40:41 cp4027 traffic_manager[20323]: [Jan 14 17:40:41.847] {0x2ade5d766700} DEBUG: <ParentRoundRobin.cc:140 (selectParent)> (parent_select) cur_index: -3, result->start_parent: -4
Jan 14 17:40:41 cp4027 traffic_manager[20323]: [Jan 14 17:40:41.847] {0x2ade5d766700} DEBUG: <ParentRoundRobin.cc:140 (selectParent)> (parent_select) cur_index: -2, result->start_parent: -4
Jan 14 17:40:41 cp4027 traffic_manager[20323]: [Jan 14 17:40:41.847] {0x2ade5d766700} DEBUG: <ParentRoundRobin.cc:140 (selectParent)> (parent_select) cur_index: -1, result->start_parent: -4
Jan 14 17:40:41 cp4027 traffic_manager[20323]: [Jan 14 17:40:41.847] {0x2ade5d766700} DEBUG: <ParentRoundRobin.cc:140 (selectParent)> (parent_select) cur_index: 0, result->start_parent: -4
```

On short-lived ATS instance, the debug output looks as expected:
```
Jan 14 18:10:41 cp5012 traffic_manager[48369]: [Jan 14 18:10:41.335] {0x2b4690202700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3120
Jan 14 18:10:41 cp5012 traffic_manager[48369]: [Jan 14 18:10:41.335] {0x2b4690202700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3121
Jan 14 18:10:43 cp5012 traffic_manager[48369]: [Jan 14 18:10:43.890] {0x2b46905e9700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3122
Jan 14 18:10:43 cp5012 traffic_manager[48369]: [Jan 14 18:10:43.890] {0x2b46905e9700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3123
Jan 14 18:10:44 cp5012 traffic_manager[48369]: [Jan 14 18:10:44.970] {0x2b46908e9700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3124
Jan 14 18:10:44 cp5012 traffic_manager[48369]: [Jan 14 18:10:44.970] {0x2b46908e9700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3125
Jan 14 18:10:45 cp5012 traffic_manager[48369]: [Jan 14 18:10:45.405] {0x2b4690a8c700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3126
Jan 14 18:10:45 cp5012 traffic_manager[48369]: [Jan 14 18:10:45.405] {0x2b4690a8c700} DEBUG: <ParentRoundRobin.cc:172 (selectParent)> (parent_select) Chosen parent = 10.132.0.112.3127
```

but ss only shows traffic hitting the odd parents:
```
vgutierrez@cp5012:~$ for port in {3120..3127}; do ss  "( dport = $port or sport = $port )" |wc -l; done
1
28
1
40
2
40
1
35
```

### PR DESCRIPTION
**Pull request by jrushford**
Fixes #6321 when proxy.config.http.no_dns_just_forward_to_parent is enabled.

When this configuration variable is enabled, a parent selection strategies findParent() function is called twice on each transaction resulting in unexpected results such as every other parent is only used in a strict round robin strategy.

### CODE REVIEW

File: proxy/ParentSelection.cc [hunk 1]
```diff
@@ -106,11 +106,13 @@ ParentConfigParams::findParent(HttpRequestData *rdata, ParentResult *result, uns
...
   ParentRecord *rec;
 
-  Debug("parent_select", "In ParentConfigParams::findParent(): parent_table: %p.", parent_table);
-  ink_assert(result->result == PARENT_UNDEFINED);
-
-  // Initialize the result structure
-  result->reset();
+  // when proxy.config.http.no_dns_just_forward_to_parent is enabled
+  // findParent() is called early in the transaction so, use that
+  // query result and clear the flag.
```
[the remark this answers is not in the document]
Comment (jvgutierrez) [reply proxy/ParentSelection.cc]: +1, IMHO ParentExists() should be an idempotent method

[the remark this answers is not in the document]
Comment (jrushford) [reply proxy/ParentSelection.cc]: @vmamidi an d @jvgutierrez I see your point.  I'll do some more work on this.

[the remark this answers is not in the document]
Comment (jrushford) [reply proxy/ParentSelection.cc]: @vmamidi and @jvgutierrez I've modified parentExists() to just search for an available parent returning true/false as is found by iterating through the parent lists.  Later in the transaction where the result is actually used, findParent() is called which looks up an available parent using the strategy algorithm.  parentExists() just needs to return true/false on whether there is an available parent and is not concerned with the selection algorithm.  parentExists() is now idempotent which fixes the strict round robin selection issue.

**Comment by zwoop** [comment]
Cherry-picked to v9.0.x branch.

**Comment by zwoop** [comment]
This has merge conflicts.

**Comment by jrushford** [comment]
@zwoop I'll have a look.

**Pull request merged**